src/cloud-resources/src/crd/materialize.rs

```diff
@@ -0,0 +1,381 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::{Condition, OwnerReference, Time};
+use kube::{CustomResource, Resource, ResourceExt};
+use schemars::JsonSchema;
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+pub const LAST_KNOWN_ACTIVE_GENERATION_ANNOTATION: &str =
+    "materialize.cloud/last-known-active-generation";
+
+pub mod v1alpha1 {
+    use kube::api::ObjectMeta;
+
+    use super::*;
+
+    #[derive(
+        CustomResource, Clone, Debug, Default, PartialEq, Eq, Deserialize, Serialize, JsonSchema,
+    )]
+    #[serde(rename_all = "camelCase")]
+    #[kube(
+        group = "materialize.cloud",
+        version = "v1alpha1",
+        kind = "Materialize",
+        singular = "materialize",
+        plural = "materializes",
+        shortname = "mzs",
+        status = "MaterializeStatus",
+        printcolumn = r#"{"name": "Organization", "type": "string", "description": "Organization ID.", "jsonPath": ".spec.organizationId", "priority": 1}"#,
+        printcolumn = r#"{"name": "ImageRef", "type": "string", "description": "Reference to the Docker image.", "jsonPath": ".spec.imageRef", "priority": 1}"#,
+        printcolumn = r#"{"name": "UpToDate", "type": "string", "description": "Whether the spec has been applied", "jsonPath": ".status.conditions[?(@.type==\"UpToDate\")].status", "priority": 1}"#
+    )]
+    pub struct MaterializeSpec {
+        pub organization_id: Uuid,
+        pub environmentd_image_ref: String,
+        pub environmentd_extra_args: Option<Vec<String>>,
+        pub external_id_prefix: Uuid,
+        pub hostname: String,
+        pub environmentd_cpu_allocation: Option<String>,
+        pub environmentd_memory_allocation: Option<String>,
+        pub request_rollout: Uuid,
+        pub force_rollout: Uuid,
+        pub in_place_rollout: bool,
+    }
+
+    impl Materialize {
+        pub fn cockroach_secret_name() -> String {
+            "cockroach-password".to_string()
+        }
+
+        pub fn cockroach_database_name(&self) -> String {
+            format!(
+                "environment_{}{}",
```

Comment:
🗒️ flagging this one as an area to discuss going forward (naming feels specific to our cloud setup)

Reply:
yeah, this ties into our existing questions around whether we want the database itself to be provisioned externally and passed in here